### PR TITLE
Hermes to periodically check for registered deployments and deploy the appscalesensor app.

### PR DESCRIPTION
**Pull request by tmarballi**
https://ocd.appscale.com:8080/job/Daily%20Build/1320/

### CODE REVIEW

File: Hermes/hermes.py [hunk 1]
```diff
@@ -1,31 +1,60 @@
...
+sys.path.append(os.path.join(os.path.dirname(__file__), '../AppServer'))
+from google.appengine.api.appcontroller_client import AppControllerException
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "../lib/"))
+import appscale_info
 
 # Tornado web server options.
 define("port", default=hermes_constants.HERMES_PORT, type=int)
 
+# The port that the SOAP server listens to.
```
Comment (menivaitsi) [inline Hermes/hermes.py]: Please move these constants in hermes_constants.

File: Hermes/hermes.py [hunk 2]
```diff
@@ -117,6 +146,94 @@ def send_all_stats():
...
+  # If the appscalesensor app is already running, then do nothing.
+  is_app_enabled = uaserver.is_app_enabled(APPSCALE_SENSOR,
+    appscale_info.get_secret())
+  if is_app_enabled == "true":
+    logging.debug("{} application is already running.".format(APPSCALE_SENSOR))
+    return
+
+  pwd = encrypt_password(USER_EMAIL, random_password_generator())
+  if create_appscale_user(pwd, uaserver) and create_xmpp_user(pwd, uaserver):
+    logging.info("Created new user and now tarring app to be deployed.")
```
Comment (menivaitsi) [inline Hermes/hermes.py]: This message should be debug.

File: Hermes/hermes.py [hunk 3]
```diff
@@ -117,6 +146,94 @@ def send_all_stats():
...
+    file_path = os.path.join(os.path.dirname(__file__), '../Apps/sensor')
+    file_suffix = 'tar.gz'
```
Comment (menivaitsi) [inline Hermes/hermes.py]: Make this a constant in hermes_constants.

File: Hermes/hermes.py [hunk 4]
```diff
@@ -117,6 +146,94 @@ def send_all_stats():
...
+      APPSCALE_SENSOR)
+    archive = tarfile.open(app_dir_location, "w|gz")
+    archive.add(file_path, arcname= APPSCALE_SENSOR)
+    archive.close()
+
+    try:
+      logging.info("Deploying the sensor app for registered deployments.")
+      acc = appscale_info.get_appcontroller_client()
+      acc.upload_app(app_dir_location, file_suffix, USER_EMAIL)
+
```
Comment (menivaitsi) [inline Hermes/hermes.py]: Remove empty line here.

File: Hermes/hermes.py [hunk 5]
```diff
@@ -117,6 +146,94 @@ def send_all_stats():
...
+        "appscalesensor app.")
+  else:
+    logging.error("Error while creating or accessing the user to deploy "
+      "appscalesensor app.")
+
+def create_appscale_user(password, uaserver):
+  """ Creates the user account with the email address and password provided. """
+  does_user_exist = uaserver.does_user_exist(USER_EMAIL, appscale_info.get_secret())
+  if does_user_exist == "true":
+    logging.info("User {0} already exists, so not creating it again.".
```
Comment (menivaitsi) [inline Hermes/hermes.py]: This message should be debug.

File: Hermes/hermes.py [hunk 6]
```diff
@@ -117,6 +146,94 @@ def send_all_stats():
...
+      format(USER_EMAIL))
+    return True
+  else:
+    if uaserver.commit_new_user(
+      USER_EMAIL, password, ACCOUNT_TYPE, appscale_info.get_secret()) == "true":
```
Comment (menivaitsi) [inline Hermes/hermes.py]: Use 4-space indentation for continuation lines.

Comment (menivaitsi) [inline Hermes/hermes.py]: Whenever the next block is indented that is.

File: Hermes/hermes.py [hunk 7]
```diff
@@ -117,6 +146,94 @@ def send_all_stats():
...
+  else:
+    if uaserver.commit_new_user(
+      xmpp_user, xmpp_pass, ACCOUNT_TYPE, appscale_info.get_secret()) == "true":
+      logging.info("XMPP username is {0}".format(xmpp_user))
+      return True
+    else:
+      logging.error("Error while creating an XMPP user.")
+      return False
+
+def random_password_generator():
```
Comment (menivaitsi) [inline Hermes/hermes.py]: This should be moved somewhere else. Probably in appscale/lib. If there's not utils file you can make one.

File: Hermes/hermes.py [hunk 8]
```diff
@@ -117,6 +146,94 @@ def send_all_stats():
...
+  """ Generates a random six character password with letters and digits. """
+  characters = string.letters + string.digits
+  pwdSize = PASSWORD_SIZE
+  return ''.join((random.choice(characters)) for x in range(pwdSize))
+
+def encrypt_password(username, password):
```
Comment (menivaitsi) [inline Hermes/hermes.py]: Same for this function.

File: Hermes/hermes.py [hunk 9]
```diff
@@ -117,6 +146,94 @@ def send_all_stats():
...
+      logging.info("XMPP username is {0}".format(xmpp_user))
+      return True
+    else:
+      logging.error("Error while creating an XMPP user.")
+      return False
+
+def random_password_generator():
+  """ Generates a random six character password with letters and digits. """
+  characters = string.letters + string.digits
+  pwdSize = PASSWORD_SIZE
```
Comment (menivaitsi) [inline Hermes/hermes.py]: Change to pwd_size.

File: Hermes/test/unit/test_hermes.py [hunk 10]
```diff
@@ -9,8 +9,20 @@
...
+import appscale_info
+
+sys.path.append(os.path.join(os.path.dirname(__file__), '../../../AppServer'))
+from google.appengine.api.appcontroller_client import AppControllerClient
+
 sys.path.append(os.path.join(os.path.dirname(__file__), "../../"))
 import helper
+import hermes
+
+import SOAPpy
```
Comment (menivaitsi) [inline Hermes/test/unit/test_hermes.py]: Move this two up on the first block of imports; maintain alphabetical order.

**Comment by tmarballi** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/1337/

File: lib/appscale_utils.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+import random
+import string
+
+def random_password_generator():
+  """ Generates a random six character password with letters and digits. """
+  characters = string.letters + string.digits
+  pwd_size = constants.PASSWORD_SIZE
+  return ''.join((random.choice(characters)) for x in range(pwd_size))
+
+def encrypt_password(username, password):
```
Comment (menivaitsi) [inline lib/appscale_utils.py]: Please add Args and Returns sections in this function.

**Comment by tmarballi** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/1341/

**Pull request merged**